src/client/Client.cc

```diff
@@ -10983,6 +10983,12 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl,
   if (movepos) {
     unlock_fh_pos(f);
   }
+  if (onfinish != nullptr) {
```

Comment:
I suggest that we use one of `BOOST_SCOPE_EXIT_*` macros to automatically complete the context (invoke callback) when a particular API returns. We would have to _always_ return 0 for async callers though, but, this would avoid the repetitive code as below.

Reply:
there are two things on my mind:

1) handle this in `Client::ll_preadv_pwritev` (the async io api) itself
2) the issue with (1) is that the internal functions `Client::_read` and `Client::_write` accept context as one of the args and so I feel context completion should be addressed in those internal functions __but__ addressing in the async api will reduce a lot of code and complexity

Reply:
handling in internal functions also implies we need to do the same use of `BOOST_SCOPE_EXIT_*` in `Client::_preadv_pwritev_locked` (the flow is `ll_preadv_pwritev` -> `_preadv_pwritev_locked` -> `_read` | `_write`)

Reply:
> there are two things on my mind:
> 
> 1. handle this in `Client::ll_preadv_pwritev` (the async io api) itself
> 2. the issue with (1) is that the internal functions `Client::_read` and `Client::_write` accept context as one of the args and so I feel context completion should be addressed in those internal functions **but** addressing in the async api will reduce a lot of code and complexity

That's correct. This is the case when a function is returning due to some error checks. Once a async call is made, the callback will be completed when the call finishes.

Reply:
so you prefer 1 or 2?

Reply:
OK. So before answering that the way I would approach this is to make the internal function (_read/_write) return a tuple - first would be a retval (0 = success, <0 on error), and second would be a flag that would signify if an async call is in-progress. That way the caller function can choose based on

- r=0, flag=true : async call in progress, callback would be automatically invoked
- r=0, flag=false : sync call (return r)
- r<0,flag=true : should have been an async call, but some checks failed - callback needs to be explicitly completed with `r`, but function would return 0.
- r<0,flag=false : sync call, some checks failed return `r`.

Reply:
do we really need a flag? the context if not nullptr is sufficient enough to assert that it's an async call

Reply:
> do we really need a flag? the context if not nullptr is sufficient enough to assert that it's an async call

nope we don't - I just used it to explain the 4 cases that needs to be covered. You can use a helper that finishes the ctx and returns either 0 (for async case) or the actual r (sync case)

Then you can do:

```
int maybe_finish_context(int r, Context *ctx) {
    if (ctx) {
        // async call
        ctx->complete(r);
        return 0;
    } else {
        // sync case
        return r;
    }
}

... top_most_funciton() {

   return maybe_finish_context(r, on_finish);
}
```

Reply:
I think introducing something looks like a C macro is really weird.  Can you use a scope guard?  Do you really need it at all?